### PR TITLE
feat!: adding BaseViewController and ViewModel

### PR DESCRIPTION
**Pull request by robinphillips**
# Adding base view controller to reduce duplication

There were some repetitive blocks of code and protocol conformances across numerous screen patterns that were creating a risk of defect and additional work should a change be required.

This PR creates a `BaseViewController` that inherits from `UIViewController`. The `BaseViewController` includes all the repetitive code and code that all screens should have. 

All screens then inherit from the `BaseViewController` instead of `UIViewController`. 

**To use:**

For most screens, in the future we should **not** need to add the following properties and methods to view model protocols:

* var rightBarButtonTitle: GDSLocalisedString? { get }
* var backButtonIsHidden: Bool { get }
* func didAppear()
* func didDismiss()

Instead, when creating concrete implementations of your view models, you should conform the concrete implementations to your view model protocol and to the `BaseViewModel` protocol. `BaseViewModel` includes the above properties and methods.

As long as your view controller inherits from `BaseViewController` instead of `UIViewController`, handling of these will be dealt with without any additional effort or code.

**Breaking change / migration**
To move to this new implementation, you should add the protocol conformance to `BaseViewModel` to concrete screen view model implementations where prompted. If there are any properties that are missing these should be added to conform to the protocol and the properties set accordingly.

If there are other screens within separate repos, it may be beneficial to inherit from `BaseViewController` and `BaseViewModel` to help reduce code duplication, but the time frame for rolling that out would be independent from the breaking changes introduced here. The breaking changes are specifically relating to screens within `GDSCommon`

**Example**

_BEFORE:_
```swift
public final class ModalInfoViewController: UIViewController {
...
}

public protocol ModalInfoViewModel {
    var title: GDSLocalisedString { get }
    var body: GDSLocalisedString { get }
    var rightBarButtonTitle: GDSLocalisedString { get }

    func didAppear()
    func didDismiss()
}
```

_AFTER:_
```swift
public final class ModalInfoViewController: BaseViewController {
...
}

public protocol ModalInfoViewModel {
    var title: GDSLocalisedString { get }
    var body: GDSLocalisedString { get }
}

// concrete view model within app:
struct ConcreteModalInfoViewModel: ModalInfoViewModel, BaseViewModel {
    var title: GDSLocalisedString  = "title"
    var body: GDSLocalisedString = "body"
    var rightBarButtonTitle: GDSLocalisedString = "right bar button"

    func didAppear() { }
    func didDismiss() { }
}
```

## Before raising your pull request:
- [x] Ran the app locally ensuring it builds 
- [x] Ran the tests locally ensuring they pass on Build
- [x] Pull request has a clear title with ticket ID and a short description about the feature or update
      i.e. _DCMAW-222: Added ReadID SDK to iOS app_
- [x] Created a `draft` pull request if it is not yet ready for review

## Before your pull request can be reviewed:
- [x] Met all of the acceptance criteria specified in the user story on Jira
- [x] Reviewed your own code to ensure you are following the style guidelines
- [x] Ran the app and tested the feature on a range of device sizes
      Please include iPod Touch, iPhone SE and iPhone 11 as a minimum.
- [x] Written Unit and Integration tests if needed

- [ ] Met all accessibility requirements?
    - [ ] Checked dynamic type sizes are applied
    - [ ] Checked VoiceOver can navigate your new code
    - [ ] Checked a user can navigate only using a keyboard around your new code 

## Before merging your pull request:
- [ ] Ensure that the code coverage and SonarCloud checks have passed
- [ ] Actioned and resolved all comments, reaching out to reviewers for clarifications if necessary.
- [ ] Ran the app to ensure that no regressions have been caused by changes during code review.
- [ ] Targeted the correct branch; `develop`, `release` or `main`

### CODE REVIEW

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockDatePickerScreenViewModel.swift [hunk 1]
```diff
@@ -8,6 +8,7 @@ struct MockDatePickerScreenViewModel: DatePickerScreenViewModel {
     var datePickerFooter: String? = "Example date picker footer"
     var buttonViewModel: ButtonViewModel
     var rightBarButtonTitle: GDSLocalisedString? = "Right bar button"
+    var backButtonIsHidden: Bool = false
```
Comment (Oliver-Binns) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/MockDatePickerScreenViewModel.swift:11]: `let`?

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockDatePickerScreenViewModel.swift [hunk 2]
```diff
@@ -8,6 +8,7 @@ struct MockDatePickerScreenViewModel: DatePickerScreenViewModel {
     var datePickerFooter: String? = "Example date picker footer"
```
Comment (Oliver-Binns) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/MockDatePickerScreenViewModel.swift:8]: should these all not be `let`?

File: Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInstructions/GDSInstructionsViewModel.swift [hunk 3]
```diff
@@ -17,7 +17,7 @@ import UIKit
 /// for other code such as making an API call.
 /// If `rightBarButtonTitle` is `nil` then the navigation item right bar button is hidden
 /// Otherwise, it is set to the title property and the action is set to the `didDismiss()` action.
-public protocol GDSInstructionsViewModel {
+public protocol GDSInstructionsViewModel: BaseViewModel {
```
Comment (Oliver-Binns) [inline Sources/GDSCommon/UI/UIKit/SharedViews/Patterns/GDSInstructions/GDSInstructionsViewModel.swift:20]: I don't think we want to have all these view models automatically conform to this. I think it should be up to the developer to add it if they need it.

It's [interface segregation principle](https://en.wikipedia.org/wiki/Interface_segregation_principle). The point of having both protocols is that they're both optional.

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/MockDatePickerScreenViewModel.swift (hunk 1 above)
Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/MockDatePickerScreenViewModel.swift:11]: I could change it but it feels like outside the scope here

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift [hunk 4]
```diff
@@ -85,7 +85,9 @@ enum Screens: String, CaseIterable {
             view.isModalInPresentation = true
             return view
         case .gdsListOptions:
-            return ListOptionsViewController(viewModel: MockListViewModel())
+            return ListOptionsViewController(viewModel: MockListViewModel(dismissAction: {
+                navigationController.popToRootViewController(animated: true)
```
Comment (benmckillop) [inline GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift:89]: Can we have this as a trailing closure and save a set of brackets? Maybe not possible if its within the vc brackets

**Review by benmckillop — APPROVED** [review]

File: GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift (hunk 4 above)
Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-Demo/Mocks/Screens.swift]: I've resolved this (partially anyway) by moving the complexity into the view controller initialisation extension file.

File: GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        viewModel = nil
+        sut = nil
+        
+        super.tearDown()
+    }
+}
+
+extension BaseViewControllerTests {
+    func test_labelContents() throws {
+        dump(sut)
```
Comment (Oliver-Binns) [inline GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift:29]: do we need this `dump`?

File: Sources/GDSCommon/README.md [hunk 6]
```diff
@@ -122,6 +122,84 @@ loadResourceWithCallback() {
...
 
+### BaseViewController
+
+``BaseViewController`` inherits from `UIViewController`. ``BaseViewController`` includes the repetitive code that all screens should have to avoid repetition and reduce risk of missing important functionality and to make it easier to amend, or fix defects if they arise.
+
+Screen view controllers should generally inherit from ``BaseViewController`` instead of `UIViewController` unless the functionality of the screen needs to be intentionally different from standard screens.
+
+#### To use:
+
+Once inheriting from ``BaseViewController`` and adding conformance to ``BaseViewModel`` to concrete view model, in the future do not add the following properties and methods to view model protocols:
```
Comment (Oliver-Binns) [inline Sources/GDSCommon/README.md:133]: It feels like this documentation is fulfilling two purposes:
1. Explaining how the package is built to our internal developers
2. Explaining how to use the package to the consumers of the package

I think the `README` should focus on the latter, and we should have additional documentation for our internal folks, as needed.

**Comment by Oliver-Binns** [comment]
I'm happy with all the proposed changes, just have the questions above around how we write our documentation - I think the versioning and upgrade docs is probably really key for us to start getting right so we don't disrupt all our teams when a breaking change gets merged.

**Comment by robinphillips** [comment]
> I'm happy with all the proposed changes, just have the questions above around how we write our documentation - I think the versioning and upgrade docs is probably really key for us to start getting right so we don't disrupt all our teams when a breaking change gets merged.

Where would you see those versioning and upgrade docs living?

File: GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift (hunk 5 above)
Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift:29]: Nope! Thanks for spotting that :)

File: GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+    
+    override func tearDown() {
+        viewModel = nil
+        sut = nil
+        
+        super.tearDown()
+    }
+}
+
+extension BaseViewControllerTests {
```
Comment (Oliver-Binns) [inline GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift:27]: should we also have tests for the actions?

Comment (robinphillips) [reply GDSCommon-Demo/GDSCommon-DemoTests/BaseViewControllerTests.swift:33]: tests added

**Review by Oliver-Binns — APPROVED** [review]

**Pull request merged**